### ISSUE TITLE
Nagios blocks on scheduling log rotation one hour before time change [relates to Daylight Saving Time]

### ISSUE THREAD

**Issue by Coservit** [NagiosEnterprises/nagioscore#626]
Hello,

we've tested Nagios behavior in the wake of next time change (-> DST) on 31/03/2019 (CEST time). During simulation of several scenarios to see if the time change doesn't cause service interruption, we remarked that Nagios behaves incorrectly during one hour preceding time change. 

Summary:
=======
During one hour before time change from NON-DST to DST (2:00 to 3:00 time change) we observe that Nagios blocks main thread on hourly log rotation scheduling (between 1 and 2 o'clock) causing big increase in event spamming and 1000 times increase in communication between ndomod.o and ndo2db server.

Environment:
=========
Ubuntu 16.04 (x86_64)
Nagios 4.2.3
Log rotate: h (hourly) - log_rotation_method=h
Hosts: 158
Services:  222

Scenario:
======
1) Stop Nagios, remove all save status files and clean all progress files
2) stop time synchronization (service ntp stop)
3) set the time to 31/03/2019 01:00:00 CEST (UTC+1)
4) confirm time (date)
5) start Nagios (service nagios start) 

Observation:
========
1) After the step 5 Nagios starts thinking that it is 31/03/2019 and initializes correctly
2) after initialization it goes to the loop of scheduling (hourly) log rotation
3) in the debug log (nagios.debug) we can see  ONLY one set of messages in loop:

[1553990405.565874] [12288.1] [pid=12688] ## Polling 0ms; sockets=6; events=387; iobs=0x19ea140
[1553990405.567789] [001.0] [pid=12688] handle_timed_event() start
[1553990405.567818] [008.0] [pid=12688] ** Timed Event ** Type: EVENT_LOG_ROTATION, Run Time: Sun Mar 31 01:00:00 2019
[1553990405.567822] [008.0] [pid=12688] ** Log File Rotation Event. Latency: 0.568s
[1553990405.567829] [064.1] [pid=12688] Making callbacks (type 11)...
[1553990405.567859] [001.0] [pid=12688] handle_timed_event() end
[1553990405.567865] [001.0] [pid=12688] reschedule_event()
[1553990405.567872] [001.0] [pid=12688] add_event()

4) these messages loop from 01:00:00 until 02:00:00 (during one hour). Nothing else is logged in debug.log - no service checks, no host checks, not any other activity. These are logged correctly otherwise (when outside this time frame)
5) we can see that checks are processed anyway, by workers on another threads, but they don't appear in nagios.debug log.
6)  We use broker ndomod.o that sends events to ndo2db server on central machine and during this hour we observe augmentation of network rate between ndomod.o and ndo2db of 1000 times (we move from 1.2 kbps to 1.2 Mbps)

Code Analysis:
==========
We've analysed the code and found out that there is an error when calculation of next log rotation time in function get_next_log_rotation_time(void) in base/utils.c. The function calculates the next hourly log rotation time "correctly" but tries to adapt the result by substracting 3600 seconds (1 hour) incorrectly.
Please see attached testing source file to see the behaviour of this function during the problem period (..._problem.c) and second source file for proposed fix (..._ok.c). I can prepare PR with proposed solution.
[get_next_log_rotation_time_test.zip](https://github.com/NagiosEnterprises/nagioscore/files/2992704/get_next_log_rotation_time_test.zip)

When launched with the original/problematic version of the function we see that the time stamp generated are not one hour ahead, but on the hour passed (same timestamp):

gcc get_next_log_rotation_time_problem.c
date -s '2019-03-31 01:00:00';./a.out
Sun Mar 31 01:00:00 CET 2019
1553990400 -> 1553990400

When launched with corrected version of the function we can see that the timestamps difference is one hour (3600 seconds):

gcc get_next_log_rotation_time_ok.c
date -s '2019-03-31 01:00:00';./a.out
Sun Mar 31 01:00:00 CET 2019
1553990400 -> 1553994000

Just for the reference, there is a similar situation with this function and the other time change (27/10/2019 02:00:00 - DST -> NON-DST) - we need to use 01:59:59 because otherwise system thinks that we are already after the time change. We expect to see the 1 second difference.
Problematic function:

gcc get_next_log_rotation_time_problem.c
date -s '2019-10-27 01:59:59';./a.out
Sun Oct 27 01:59:59 CEST 2019
1572134399 -> 1572134400

we have one second difference instead of correct 1 hour and one second difference. That we get when launched the corrected version:

gcc get_next_log_rotation_time_ok.c
date -s '2019-10-27 01:59:59';./a.out
Sun Oct 27 01:59:59 CEST 2019
1572134399 -> 1572138000

difference 3601 seconds - what we expect.

Correction description:
===============
the field tm_isdst should be set to -1 before calling mktime(...). That will assure that mktime will try to handle dst properly (see man mktime)
we should remove the code that tries to compensate the one hour shift during dst change.

#if 0//!COSERVIT
        if(is_dst_now == TRUE && t->tm_isdst == 0)
                run_time += 3600;
        else if(is_dst_now == FALSE && t->tm_isdst > 0)
                run_time -= 3600;
#endif

I checked the last code of the aforementioned function in depot git and did not find any correction. 
Please let me know if you think I missed something in my description. Feel free to ask any additional questions.

**Comment by jomann09** [comment]
I believe this was already resolved in the maint branch if you'd like to try it. There was a pull request (#610) to fix it, although not in the same way that you had mentioned as a fix. I will test this again to verify it too.

**Comment by Coservit** [comment]
Unfortunately the solution in #610 does not work for us. The time change in CEST is at 2 and 3 o'clock in the morning so the day doesn't decrement after subtracting one hour.
We should not actually do any compensation, we should really just use the localtime_r's calculation. Assuming that we give the proper information to localtime_r call. It should work in any timezone for any time change, if the OS has all necessary info.
For me, this issue is not really a Duplicate. It is related, but not resolved by the issue #610.

**Comment by jomann09** [comment]
I'll definitely test it out, and I agree that manipulating the time manually is not a good way of doing it. I'll leave this open as a bug and we will get it fixed in a coming release.

**Comment by Madlohe** [comment]
I've tried this out in a variety of times and timezones, and it seems that lines 1598-1599:
```
else if(is_dst_now == FALSE && t->tm_isdst > 0)
   run_time -= 3600; 
```
are simply never correct. @Coservit, I've pushed the fix to https://github.com/Madlohe/nagioscore/tree/bug/626/daylight-saving if you're interested in trying that build

**Comment by Madlohe** [comment]
Merged #648. Closing
